In [1]:
sc.install_pypi_package("boto3 ")
sc.install_pypi_package("pandas ")
sc.install_pypi_package("fsspec")
sc.install_pypi_package("s3fs ")

VBox()

Starting Spark application


FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

SparkSession available as 'spark'.


FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

  Using cached https://files.pythonhosted.org/packages/13/a1/d9e77245939ec608f0787545824356705f2341c395c5a37e9778bdb1cd98/boto3-1.15.0-py2.py3-none-any.whl
  Using cached https://files.pythonhosted.org/packages/22/0a/26bb66f98db0950526f1d9fd7acf46c2ec9244e8178fe2c6be2fc77f9d70/botocore-1.18.0-py2.py3-none-any.whl
  Using cached https://files.pythonhosted.org/packages/69/79/e6afb3d8b0b4e96cefbdc690f741d7dd24547ff1f94240c997a26fa908d3/s3transfer-0.3.3-py2.py3-none-any.whl
  Using cached https://files.pythonhosted.org/packages/9f/f0/a391d1463ebb1b233795cabfc0ef38d3db4442339de68f847026199e69d7/urllib3-1.25.10-py2.py3-none-any.whl
  Using cached https://files.pythonhosted.org/packages/d4/70/d60450c3dd48ef87586924207ae8907090de0b306af2bce5d134d78615cb/python_dateutil-2.8.1-py2.py3-none-any.whl

  Using cached https://files.pythonhosted.org/packages/74/69/18b96b520519818e00b04dd08d7cbc5e764f1465f5a280cf96173f34c54e/pandas-1.1.2-cp37-cp37m-manylinux1_x86_64.whl

  Using cached https://files.py

### Instalación de librerías necesarias y adecuación de ambiente

In [2]:
import pandas as pd
import boto3
import io

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Se cargan los archivos archivos desde S3 (AWS) con privacidad 'Pública'

In [3]:
myaws_access_key_id='ASIATJCFBBZFUTF3MWJO'
myaws_secret_access_key='y+G5B9GDkwRCej+ayzVHxYymEwLHmrc3cBzlFPAq'
myaws_session_token= 'FwoGZXIvYXdzEB0aDHXa4OEfpffgnpBIQyLDAeNfOATcCTgKNE6Vus11vZBoTbe8du+W2ZnbfSYqAlqvMRa/payOKgPZlITOZ3jEvCOOQPMj7SwHBOsb48JZO/grEipxtkfpdCmMJu6jWFcnB2Q8+yWETO7SWUeyPt4YlHFksCEUn/zVUp5BL8U/w17lOTR3dnbat4ankPQMP1YxbGClTB5Py6VEYb5J+Jc+yyejLaA7r1MaAQT3SyqrazaWylPlOPeYBXoJkwqNsNuKgW30sX7AxV6O3kyu5Ckm3WQ8LSjZ1ZD7BTItGc9Wo276nszjvEvM31MRkNc04N7JsO3sXSb4Kpg5gcasByU0e1EHTP204MMZ'
# Nombre del bucket
mybucket_name = 'resultados-mineria'

# Se coencta un cliente a S3.
s3client = boto3.client('s3',
    aws_access_key_id=myaws_access_key_id,
    aws_secret_access_key=myaws_secret_access_key,
    aws_session_token=myaws_session_token
)

s3resource = boto3.resource('s3',
    aws_access_key_id=myaws_access_key_id,
    aws_secret_access_key=myaws_secret_access_key,
    aws_session_token=myaws_session_token
)

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Se crean 6 funciones propias para preprocesar la informarción:

test_s3object -> Verifica la conexión a S3

create_s3folder -> Crea las carpetas en el bucket S3


list_s3dir -> Lista los archivos del directorio


write_s3file -> Crea archivos en S3


upload_s3file -> Carga archivos a S3


read_s3file -> Lee archivos de S3

In [4]:
# Se crean las funciones para acceder a S3.
def test_s3object(prefix):
    mybucket = s3resource.Bucket(mybucket_name) # just Bucket name
    obj = list(mybucket.objects.filter(Prefix=prefix))
    if len(obj) > 0:
        return True
    else:
        return False
    
def create_s3folder(folder_name):
    s3client.put_object(Bucket=mybucket_name, Key=(folder_name+'/'))
    
def list_s3dir(prefix):
    continuation_token = None
    Temp=[]
    while True:
        kwargs = {'Bucket': mybucket_name, 'Prefix': prefix}
        list_kwargs = dict(MaxKeys=1000, **kwargs)
        if continuation_token:
            list_kwargs['ContinuationToken'] = continuation_token
        response = s3client.list_objects_v2(**list_kwargs)
        temp1 = []
        temp1 = list(response.get('Contents', []))
        for i in temp1:
            Temp.append(i['Key'])
        #yield from response.get('Contents', [])
        if not response.get('IsTruncated'):  # At the end of the list?
            break
        continuation_token = response.get('NextContinuationToken')
    return Temp

'''
def list_s3dir(prefix):
    keys = []
    kwargs = {'Bucket': mybucket_name, 'Prefix': prefix}
    resp = s3client.list_objects_v2(**kwargs)
    for obj in resp['Contents']:
        keys.append(obj['Key'])
    return keys
'''

def write_s3file(fname,fcontent):
    file = s3resource.Object(mybucket_name,fname)
    file.put(Body=fcontent)
    
def upload_s3file(file_name, object_name=None):
    """Upload a file to an S3 bucket

    :param bucket: Bucket to upload to
    :param file_name: File to upload
    :param object_name: S3 object name. If not specified then file_name is used
    :return: True if file was uploaded, else False
    """

    # If S3 object_name was not specified, use file_name
    if object_name is None:
        object_name = file_name

    # Upload the file
    try:
        response = s3client.upload_file(file_name, mybucket_name, object_name)
    except ClientError as e:
        logging.error(e)
        return False
    return True

def read_s3file(fname):
    data = s3client.get_object(Bucket = mybucket_name, Key = fname)
    return data['Body']


VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

## Carga de información

In [5]:
files = 'Resultados/'
A = list_s3dir(files)
print(len(A))
A.pop(0);

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

1349
'Resultados//'

In [ ]:
files = 'Resultados/'
filenames = list_s3dir(files)
print(len(filenames))
filenames.pop(0);

In [6]:
print(len(filenames))

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

1348

### Procesamiento y recomendación de cercanos

In [7]:
resultados = pd.DataFrame(columns=['Recomendadas', 'Usuario'])
for fname in filenames:
    data = s3client.get_object(Bucket=mybucket_name, Key=fname)
    datos = pd.read_csv(io.BytesIO(data['Body'].read()))
    lista = []
    lista = []
    for i in list(datos.columns):
        canciones = i.replace("[","")
        canciones = canciones.replace("'","")
        canciones = canciones.replace("]","")
        canciones = canciones.replace(" ","")
        lista.append(canciones)
    data2 = pd.DataFrame(lista,columns=['Recomendadas'])
    user = fname[11:]
    user = user[:-4]
    data2['Usuario'] = user
    resultados = resultados.append(data2, sort=False)

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [11]:
resultados.groupby('Usuario').count()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

                                          Recomendadas
Usuario                                               
000ef25cc955ad5841c915d269432eea41f4a1a5            30
002543003041db1d049206b09426d5cdffc0f451            30
0031572620fa7f18487d3ea22935eb28410ecc4c            30
00560ea172697c2a548df4812e43a4eadae0c2e9            30
006ed26a7a1b4b245ff1be9eeedccde90e5ff285            30
...                                                ...
ff414a33b79d5908f560e538dab7d43453e560bd            30
ff549dcb42f607cba44e1986d2ef155effc1e370            30
ff908353ca57d4733782eda1bdaa5bd3a26b3bca            30
ff951aee373973ccdacf0d766e9b8c8f4fb56eb8            30
ffd73451d81c7cdeda85810b3bb2453175bd51c8            30

[1348 rows x 1 columns]

In [98]:
triplets_file = 's3://mineria-msd/small_data/x_test.csv'
test = pd.read_csv(triplets_file)
test = test.drop(['Unnamed: 0','play_count'],axis = 1)

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Se cargan los datos de entrenamiento del modelo y se seleccionan las primeras 5.000 canciones del listado de canciones reproducidas

In [99]:
from pyspark.sql import *
from pyspark.sql.functions import *
from pyspark.sql.functions import rank, col
triplets_file = 's3://mineria-msd/small_data/x_train.csv'
listened_songs_df = spark.read.format("csv").option("header", "true").load(triplets_file)
listened_songs_df = listened_songs_df.withColumnRenamed('play_count','listen_count')
listened_songs_df = listened_songs_df.withColumnRenamed('user','user_id')
listened_songs_df = listened_songs_df.withColumnRenamed('song','song_id')
listened_songs_df = listened_songs_df.filter(listened_songs_df.listen_count > 1)
listened_songs_df.groupby("song_id").count().sort(desc("count")).count()
Top = listened_songs_df.groupby("song_id").count().sort(desc("count")).limit(5000)

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Conversión a Pandas

In [100]:
Top = Top.select("*").toPandas()
Top = Top.rename(columns={"song_id": "song"})

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Se integran los datos de entrenamiento con el Top de canciones y se estiman las recomendaciones para cada una de las combinacione "usuario/canción"

In [102]:
test = pd.merge(test,Top,on='song')
test = test.drop(['count'],axis = 1)
canXusu = test.groupby('user').count().rename(columns={"song": "NumCancion"})
resultados['marca']=1
resultados = resultados.rename(columns={"Recomendadas": "song","Usuario": "user"})
result = pd.merge(test, resultados, on=['user', 'song'])
result = result.drop(['marca'],axis = 1)

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

Contamos el nuero de recomendaciones que haya acertado el modelo

In [118]:
recXusu = result.groupby('user').count().rename(columns={"song": "acierto"})

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

integramos los datos entre para validar el desempeno del módelo

In [121]:
result_user = pd.merge(recXusu,canXusu,on='user')

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

El cáclulo de la presición empleado es contando en cuántas canciones se acertó sobre el total de la población de canciones del usuario

In [126]:
result_user['precision']=result_user['acierto']/result_user['NumCancion']

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

In [127]:
result_user['precision'].mean()

VBox()

FloatProgress(value=0.0, bar_style='info', description='Progress:', layout=Layout(height='25px', width='50%'),…

0.4108804483600534

Se obtiene una presición del 41%, en sistemas de recomendación es considerada un valor relativamente alto. Desde la percepción real de estas soluciones analíticas, se ha evidenciado que modelos de recomendación de películas en plataformas de streaming no suelen ser tan precisos